IllegalType: illegal types in overridden methods should not be reported

```
/var/tmp $ javac Foo.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="IllegalType">
            <property name="illegalClassNames" value="StringBuffer"/>
        </module>
    </module>
</module>

/var/tmp $ cat Bar.java
interface Bar {
    void bar(StringBuffer buffer);
    StringBuffer bar();
    Object baz();
}

/var/tmp $ cat Foo.java
class Foo implements Bar {
    public void foo(StringBuffer buffer) { // line 2
    }
    public StringBuffer foo() { // line 4
        return null;
    }
    @Override
    public void bar(StringBuffer buffer) { // line 8
    }
    @Override
    public StringBuffer bar() { // line 11
        return null;
    }
    @Override
    public StringBuffer baz() { // line 15
        return null;
    }
}

/var/tmp $ java -Duser.language=en -Duser.country=US -jar checkstyle-8.18-all.jar -c config.xml Foo.java
Starting audit...
[ERROR] D:\Sourcecode\other\discord-logger\tmp\Foo.java:2:21: Usage of type 'StringBuffer' is not allowed. [IllegalType]
[ERROR] D:\Sourcecode\other\discord-logger\tmp\Foo.java:4:12: Usage of type 'StringBuffer' is not allowed. [IllegalType]
[ERROR] D:\Sourcecode\other\discord-logger\tmp\Foo.java:8:21: Usage of type 'StringBuffer' is not allowed. [IllegalType]
[ERROR] D:\Sourcecode\other\discord-logger\tmp\Foo.java:11:12: Usage of type 'StringBuffer' is not allowed. [IllegalType]
[ERROR] D:\Sourcecode\other\discord-logger\tmp\Foo.java:15:12: Usage of type 'StringBuffer' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 5 errors.
```

---------------

The 8:21 finding should not be there.
If you override a method, there is no chance to change the parameter types in any way or you overload the method instead.

--------------

**_Additionally demanded by project maintainers:_**
11:12 and 15:12 should not be reported either. Even though original return type is `Object` and can be changed in the overridden class as seen in the example. It will be hard to decide where to raise violation and where not to. It is better to have some false negatives rather than false positives.
I'm not sure it is possible to write good suppression for such cases.
All overridden methods should be ignored from validation.

--------------

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
@@ -31,6 +31,7 @@ import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.FullIdent;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.AnnotationUtil;
 import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
@@ -284,7 +285,7 @@ public final class IllegalTypeCheck extends AbstractCheck {
      * @param methodDef method for check.
      */
     private void visitMethodDef(DetailAST methodDef) {
-        if (isVerifiable(methodDef) && isCheckedMethod(methodDef)) {
+        if (isCheckedMethod(methodDef)) {
             checkClassName(methodDef);
         }
     }
@@ -296,9 +297,7 @@ public final class IllegalTypeCheck extends AbstractCheck {
     private void visitParameterDef(DetailAST parameterDef) {
         final DetailAST grandParentAST = parameterDef.getParent().getParent();
 
-        if (grandParentAST.getType() == TokenTypes.METHOD_DEF
-            && isCheckedMethod(grandParentAST)
-            && isVerifiable(grandParentAST)) {
+        if (grandParentAST.getType() == TokenTypes.METHOD_DEF && isCheckedMethod(grandParentAST)) {
             checkClassName(parameterDef);
         }
     }
@@ -531,7 +530,8 @@ public final class IllegalTypeCheck extends AbstractCheck {
     private boolean isCheckedMethod(DetailAST ast) {
         final String methodName =
             ast.findFirstToken(TokenTypes.IDENT).getText();
-        return !ignoredMethodNames.contains(methodName);
+        return isVerifiable(ast) && !ignoredMethodNames.contains(methodName)
+                && !AnnotationUtil.containsAnnotation(ast, "Override");
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheckTest.java
@@ -68,6 +68,8 @@ public class IllegalTypeCheckTest extends AbstractModuleTestSupport {
         final String[] expected = {
             "16:13: " + getCheckMessage(MSG_KEY, "java.util.TreeSet"),
             "17:13: " + getCheckMessage(MSG_KEY, "TreeSet"),
+            "42:14: " + getCheckMessage(MSG_KEY, "HashMap"),
+            "44:5: " + getCheckMessage(MSG_KEY, "HashMap"),
         };
 
         verify(checkConfig, getPath("InputIllegalType.java"), expected);
@@ -85,6 +87,8 @@ public class IllegalTypeCheckTest extends AbstractModuleTestSupport {
                     + "InputIllegalType.AbstractClass"),
             "16:13: " + getCheckMessage(MSG_KEY, "java.util.TreeSet"),
             "25:36: " + getCheckMessage(MSG_KEY, "java.util.TreeSet"),
+            "42:14: " + getCheckMessage(MSG_KEY, "HashMap"),
+            "44:5: " + getCheckMessage(MSG_KEY, "HashMap"),
         };
 
         verify(checkConfig, getPath("InputIllegalType.java"), expected);
@@ -98,6 +102,8 @@ public class IllegalTypeCheckTest extends AbstractModuleTestSupport {
         final String[] expected = {
             "16:13: " + getCheckMessage(MSG_KEY, "java.util.TreeSet"),
             "17:13: " + getCheckMessage(MSG_KEY, "TreeSet"),
+            "42:14: " + getCheckMessage(MSG_KEY, "HashMap"),
+            "44:5: " + getCheckMessage(MSG_KEY, "HashMap"),
         };
 
         verify(checkConfig, getPath("InputIllegalType.java"), expected);
@@ -115,6 +121,8 @@ public class IllegalTypeCheckTest extends AbstractModuleTestSupport {
                     + "InputIllegalType.AbstractClass"),
             "16:13: " + getCheckMessage(MSG_KEY, "java.util.TreeSet"),
             "17:13: " + getCheckMessage(MSG_KEY, "TreeSet"),
+            "42:14: " + getCheckMessage(MSG_KEY, "HashMap"),
+            "44:5: " + getCheckMessage(MSG_KEY, "HashMap"),
         };
 
         verify(checkConfig, getPath("InputIllegalType.java"), expected);

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalType.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalType.java
@@ -1,8 +1,8 @@
 package com.puppycrawl.tools.checkstyle.checks.coding.illegaltype;
+import java.util.HashMap;
 import java.util.TreeSet;
-import java.util.Hashtable;
 //configuration: default
-public class InputIllegalType {
+public class InputIllegalType implements InputIllegalTypeSuper {
     private AbstractClass a = null; //WARNING
     private NotAnAbstractClass b = null; /*another comment*/
 
@@ -16,11 +16,32 @@ public class InputIllegalType {
     private java.util.TreeSet table1() { return null; } //WARNING
     private TreeSet table2() { return null; } //WARNING
     static class SomeStaticClass {
-        
+
     }
-    
+
     InputIllegalType(Integer i) {}
     private void table2(Integer i) {}
 
     private void getInitialContext(java.util.TreeSet v) {} // ignore method by default
+
+    @Override
+    public void foo(HashMap<?, ?> buffer) {} // ignore
+
+    @Override
+    public HashMap<?, ?> foo() { //ignore
+        return null;
+    }
+
+    @Override
+    public HashMap<?, ?> bar() { //ignore
+        return null;
+    }
+}
+
+interface InputIllegalTypeSuper {
+    void foo(HashMap<?, ?> buffer); //WARNING
+
+    HashMap<?, ?> foo(); //WARNING
+
+    Object bar();
 }

